Dynatrace metrics exporter

dyladan opened the pull request:
**Description:** Add an exporter for metrics to Dynatrace

**Link to tracking Issue:** n/a

**Testing:** Tests are done for configuration and some serialization

**Documentation:** README.md added in `exporter/dynatraceexporter` which explains configuration and use of Dynatrace metric exporter.

dyladan committed:
Dynatrace metrics exporter

dyladan commented:
@kbrockhoff not sure who to tag here but you're assigned to this PR so I guess you're as good a person to ask as any. The tests seem to have failed in a completely unrelated part of the codebase (xray receiver). Is there a way you can force a rerun? Or should I just push another commit to trigger the build?

anuraaga commented on exporter/dynatraceexporter/serialization/serialization.go:
```diff
@@ -0,0 +1,178 @@
+
+		tagline := serializeTags(p.LabelsMap(), tags)
+		count := float64(p.Count())
+		avg := p.Sum() / count
```
Think we need to handle `count == 0`

anuraaga commented on exporter/dynatraceexporter/serialization/serialization.go:
```diff
@@ -0,0 +1,178 @@
+
+		tagline := serializeTags(p.LabelsMap(), tags)
+
+		avg := float64(p.Sum()) / float64(p.Count())
```
Ditto

anuraaga commented on exporter/dynatraceexporter/serialization/serialization.go:
```diff
@@ -0,0 +1,178 @@
+		if err != nil {
+			return
+		}
+		tag := key + "=" + v
```
Assuming it's safe to not have any escaping or anything of the value?

dyladan replied:
Added escaping but I think the characters requiring escaping are not allowed by the spec anyways

dyladan commented on exporter/dynatraceexporter/metrics_exporter.go:
```diff
@@ -0,0 +1,186 @@
+
+	req.Header.Add("Content-Type", "text/plain; charset=UTF-8")
+
+	if len(e.cfg.API.APIToken) != 0 {
```
I moved the validation while developing. you are correct it's validated

dyladan commented:
@anuraaga I addressed all of your comments except I am still building the string. I will try to get around to refactoring it to use a buffer but i don't know when that will be done and I dont think it should block this PR.

dyladan commented on exporter/dynatraceexporter/metrics_exporter.go:
```diff
@@ -0,0 +1,186 @@
+
+func (e *exporter) send(message string) error {
+	e.logger.Debug("Sending lines to Dynatrace\n" + message)
+	req, err := http.NewRequest("POST", e.cfg.API.URL, bytes.NewBufferString(message))
```
I put this on my todo list but I don't think it should block this PR.

bogdandrutu commented on exporter/dynatraceexporter/LICENSE:
```diff
@@ -0,0 +1,201 @@
+                           Version 2.0, January 2004
+                        http://www.apache.org/licenses/
+
+   TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION
```
Why is this needed?

dyladan replied:
I'll remove it, it was added by my client and I didn't notice.

bogdandrutu commented on exporter/dynatraceexporter/config/config.go:
```diff
@@ -0,0 +1,58 @@
+type Config struct {
+	configmodels.ExporterSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct.
+
+	// URL of the Dynatrace metrics ingest endpoint. Defaults to http://127.0.0.1:14499/metrics/ingest
+	Endpoint string `mapstructure:"endpoint"`
```
Do you want to use the confighttp.HttpClientSettigs from core?

dyladan replied:
I wasn't aware it existed. I just looked at another exporter and did what it did. I think it was datadog? I can change to use that if you would prefer?

bogdandrutu reviewed (commented):
Please follow the instructions https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#when-adding-a-new-component

kbrockhoff commented on exporter/dynatraceexporter/factory.go:
```diff
@@ -0,0 +1,78 @@
+		params.Logger,
+		exp.PushMetricsData,
+		exporterhelper.WithQueue(exporterhelper.CreateDefaultQueueSettings()),
+		exporterhelper.WithRetry(exporterhelper.CreateDefaultRetrySettings()),
```
Do you have plans to make these configurable in the future?

dyladan replied:
No. This is just what I saw another exporter do.

kbrockhoff replied:
This is fine to start but as this gets deployed to different environments you will likely find this needs to be adjustable.

kbrockhoff commented on exporter/dynatraceexporter/metrics_exporter.go:
```diff
@@ -0,0 +1,181 @@
+				continue
+			}
+			metrics := libraryMetric.Metrics()
+			for k := 0; k < metrics.Len(); k++ {
```
Consider handling processing for loop in a separate method.

dyladan replied:
All that is done in the loop is some nil checking then it delegates to the correct serialization function. I could make the whole for loop body its own function but I think it unnecessary redirection while reading the code.

dyladan commented:
> Please follow the instructions [open-telemetry/opentelemetry-collector@`master`/CONTRIBUTING.md#when-adding-a-new-component](https://github.com/open-telemetry/opentelemetry-collector/blob/master/CONTRIBUTING.md#when-adding-a-new-component)

So would you like me to open a new PR with just the Readme, configuration, and factory?

dyladan commented:
What is the best way for me to resolve go.sum issues? Should I just re-run `go build` or should I use `go mod tidy`? `go mod tidy` removes the `memcachedreceiver` also for some reason.

dyladan committed:
Merge remote-tracking branch 'origin/master' into dynatrace-exporter

dyladan commented:
I just wanted to check in to see what the current status of this is. I don't think this PR is waiting on me to do something but if it is please let me know. I'm not sure how many reviews are needed for this repo before a merge, but I think I've addressed all of the open PR comments. I have also removed the component from the components.go file as suggested by bogdan to limit the impact of merging this component until its stability is proven.

tigrannajaryan commented:
@bogdandrutu do you have any other comments or good to merge?
